implementations/micrometer-registry-timestream2/src/main/java/io/micrometer/timestream2/TimeStreamMeterRegistry.java

```diff
@@ -0,0 +1,343 @@
+
+    private static final WarnThenDebugLogger warnThenDebugLogger = new WarnThenDebugLogger(TimeStreamMeterRegistry.class);
+
+    private final Logger logger = LoggerFactory.getLogger(TimeStreamMeterRegistry.class);
+
+    private final TimeStreamConfig config;
+
+    private final TimestreamWriteAsyncClient timestreamWriteClient;
+
+    private volatile boolean checkedForTable = false;
+
+
+    /**
+     * Create a new instance with given parameters.
+     *
+     * @param config configuration to use
+     * @param clock  clock to use
+     * @param client TimeStream client to use
+     * @since 1.6.0
+     */
+    protected TimeStreamMeterRegistry(TimeStreamConfig config, Clock clock, TimestreamWriteAsyncClient client) {
+        this(config, clock, DEFAULT_THREAD_FACTORY, client);
+    }
+
+    /**
+     * Create a new instance with given parameters.
+     *
+     * @param config        configuration to use
+     * @param clock         clock to use
+     * @param threadFactory thread factory to use
+     * @param client        TimeStream client to use
+     * @since 1.6.0
+     */
+    protected TimeStreamMeterRegistry(TimeStreamConfig config, Clock clock, ThreadFactory threadFactory, TimestreamWriteAsyncClient client) {
+        super(config, clock);
+        config().namingConvention(new TimeStreamNamingConvention());
+        this.config = config;
+        checkClient(client);
+        this.timestreamWriteClient = client;
+        start(threadFactory);
+    }
+
+    public static Builder builder(TimeStreamConfig config) {
+        return new Builder(config);
+    }
+
+    private void checkClient(TimestreamWriteAsyncClient client) {
+        Objects.requireNonNull(client);
+    }
+
+    @Override
+    protected void publish() {
+        for (List<Meter> partition : MeterPartition.partition(this, config.batchSize())) {
+            sendRecords(metricData(partition));
+        }
+    }
+
+    //VisibleForTesting
+    List<Record> metricData(List<Meter> partition) {
+        Batch batch = new Batch();
+        return partition.stream().flatMap(m -> m.match(
+                batch::gaugeData,
+                batch::counterData,
+                batch::timerData,
+                batch::summaryData,
+                batch::longTaskTimerData,
+                batch::timeGaugeData,
+                batch::functionCounterData,
+                batch::functionTimerData,
+                batch::metricData)
+        ).collect(toList());
+    }
+
+
+    void sendRecords(List<Record> records) {
+        WriteRecordsRequest.Builder writeRecordsRequestBuilder = WriteRecordsRequest.builder()
+                .databaseName(config.database())
+                .tableName(config.table())
+                .records(records);
+        WriteRecordsRequest writeRecordsRequest = writeRecordsRequestBuilder.build();
+
+        timestreamWriteClient.writeRecords(writeRecordsRequest).whenCompleteAsync((response, t) -> {
+            if (t != null) {
```

Comment:
`CompletableFuture` has methods for normal and exceptional completion (`thenRun` and `exceptionally`, they also have async versions), this might be a little simpler using them.

Reply:
done